### DISCUSSION TITLE
Does semantic-kernel work with local llms?

### DISCUSSION THREAD

**Discussion by GeorgeDittmar** [microsoft/semantic-kernel#1256] [Q&A] [answered]
All the examples I have seen so far seem to focus on tying into  a service such as OpenAI or Hugging face but what if I have a custom llm trained locally I want to use. Is there a way to do that?

**Comment by xbotter** [comment, marked as answer]
Yes, you only need to implement the interfaces ITextCompletion, IChatCompletion, and IEmbeddingGeneration. You can refer to [Example16_CustomLLM.](https://github.com/microsoft/semantic-kernel/blob/main/samples/dotnet/kernel-syntax-examples/Example16_CustomLLM.cs)  for more information.

I am trying to use LLamaSharp to implement the Completion and Embedding interfaces of the semantic kernel. However, the performance and effectiveness of the Local LLM are not satisfactory, and there are still many requirements that need to be adapted.

https://github.com/xbotter/semantic-kernel-LLamaSharp

**Comment by jamesburton** [comment]
This is the current CustomLLM source link:
https://github.com/microsoft/semantic-kernel/blob/main/dotnet/samples/KernelSyntaxExamples/Example16_CustomLLM.cs

**Comment by jphgardner** [comment]
Is there an updated link for this?

**Reply by evchaki** [reply]
Here is an example for local models --> https://github.com/microsoft/semantic-kernel/blob/main/dotnet/samples/Concepts/LocalModels/MultipleProviders_ChatCompletion.cs and custom models --> https://github.com/microsoft/semantic-kernel/blob/bf42feda6d90e6d0fd889f1f1141c6d793eb6253/dotnet/samples/Concepts/TextGeneration/Custom_TextGenerationService.cs#L25
